dashboard/app/controllers/plc/enrollment_evaluations_controller.rb

```diff
@@ -1,25 +1,44 @@
+  before_action :authorize_and_check_unit_state
+
   def perform_evaluation
-    authorize! :read, Plc::UserCourseEnrollment
     plc_unit_assignment = Plc::EnrollmentUnitAssignment.find(params[:unit_assignment_id])
 
-    if plc_unit_assignment.status != Plc::EnrollmentUnitAssignment::PENDING_EVALUATION
-      raise "Cannot perform evaluation - unit assignment is in state #{plc_unit_assignment.status}"
-    end
-
     @questions = plc_unit_assignment.plc_course_unit.plc_evaluation_questions
     @course_unit = plc_unit_assignment.plc_course_unit
   end
 
   def submit_evaluation
-    authorize! :read, Plc::UserCourseEnrollment
-    question_responses = params[:answerModuleList].split(',')
+    # Eventually, the logic for more complex module assignments will go here. For now since we are naively assigning modules, just go to preview
+    redirect_to controller: :enrollment_evaluations, action: :preview_assignments, unit_assignment_id: params[:unit_assignment_id], enrolled_modules: params[:answer_module_list]
+  end
+
+  def preview_assignments
+    @enrollment_unit_assignment = Plc::EnrollmentUnitAssignment.find(params[:unit_assignment_id])
+    @enrolled_modules = Plc::LearningModule.where(id: params[:enrolled_modules].split(','))
+
+    if @enrolled_modules.where.not(plc_course_unit: @enrollment_unit_assignment.plc_course_unit).any?
+      redirect_to controller: :enrollment_evaluations, action: :perform_evaluation, unit_assignment_id: @enrollment_unit_assignment.id
+    end
+  end
+
+  def confirm_assignments
+    modules_to_enroll_in = Plc::LearningModule.where(id: params[:learning_module_ids])
     enrollment_unit_assignment = Plc::EnrollmentUnitAssignment.find(params[:unit_assignment_id])
 
-    modules_to_enroll_in = Plc::LearningModule.where(id: question_responses)
+    if modules_to_enroll_in.where.not(plc_course_unit: enrollment_unit_assignment.plc_course_unit).any?
+      redirect_to controller: :enrollment_evaluations, action: :perform_evaluation, unit_assignment_id: enrollment_unit_assignment.id
+    else
+      enrollment_unit_assignment.enroll_user_in_unit_with_learning_modules(modules_to_enroll_in)
+      redirect_to controller: :enrollment_unit_assignments, action: :show, id: enrollment_unit_assignment.id
+    end
+  end
+
+  private
+  def authorize_and_check_unit_state
+    plc_unit_assignment = Plc::EnrollmentUnitAssignment.find(params[:unit_assignment_id])
+    authorize! :read, plc_unit_assignment.plc_user_course_enrollment
 
-    enrollment_unit_assignment.enroll_user_in_unit_with_learning_modules(modules_to_enroll_in)
-    enrollment_unit_assignment.update(status: Plc::EnrollmentUnitAssignment::IN_PROGRESS)
-    redirect_to controller: :enrollment_unit_assignments, action: :show, id: enrollment_unit_assignment.id
+    raise 'Cannot do evaluation actions when unit is in progress' unless Plc::EnrollmentUnitAssignment::PENDING_EVALUATION == plc_unit_assignment.status
```

Comment:
When you first confirmed your module selections, then the unit was marked as in progress. Going back doesn't undo that, so yeah you should get this error.

Right now, we don't want users to be able to retake the evaluation or change their module selection once they've picked their units. Future iterations on this may allow you to change your enrollment if you haven't started task completion work